siphon/ncss.py

```diff
@@ -0,0 +1,391 @@
+import xml.etree.ElementTree as ET
+from io import BytesIO
+
+import numpy as np
+
+from .http_util import DataQuery, HTTPEndPoint, parse_iso_date
+from .ncss_dataset import NCSSDataset
+
+
+def default_unit_handler(data, units=None):  # pylint:disable=unused-argument
+    r'Default unit handler, which ignores units and just returns ``numpy.array``'
+    return np.array(data)
+
+
+class NCSS(HTTPEndPoint):
+    r'''Wraps access to the NetCDF Subset Service (NCSS) on a THREDDS server.
+
+    Simplifies access via HTTP to the NCSS endpoint. Parses the metadata, provides
+    data download and parsing based on the appropriate query.
+
+    Attributes
+    ----------
+    metadata : ``NCSSDataset`` instance
+        Contains the result of parsing the NCSS endpoint's dataset.xml. This has
+        information about the time and space coverage, as well as full information
+        about all of the variables.
+    variables : set of strings
+        Names of all variables available in this dataset
+    unit_handler : callable
+        Function to handle units that come with CSV/XML data. Should be a callable that
+        takes a list of string values and unit str (can be ``None``), and returns the
+        desired representation of values. Defaults to ignoring units and returning
+        ``numpy.array``.
+    '''
+
+    # Need staticmethod to keep this from becoming a bound method, where self
+    # is passed implicitly
+    unit_handler = staticmethod(default_unit_handler)
+
+    def _get_metadata(self):
+        # Need to use .content here to avoid decode problems
+        meta_xml = self.get_path('dataset.xml').content
+        root = ET.fromstring(meta_xml)
+        self.metadata = NCSSDataset(root)
+        self.variables = set(self.metadata.variables.keys())
```

Comment:
The motivation of `set` wasn't duplicates, but sub-linear (I think it's a hash, like `dict`, so O(1)) search. If we want duplicate checking, I agree that it should be in `NCSSDataset`.